### PR TITLE
Added configurable immutable types whitelist

### PR DESCRIPTION
**Pull request by mwelsh1118**
To address issue #31, I first refactored the test support code to use composition rather than inheritance based on Microsoft's Roslyn tests. The configurable whitelist is provided via an additional file on the analyzed project. Added implicitly to the whitelist are the hard-coded types System.Guid/System.TimeSpan/etc.

### CODE REVIEW

**Comment by dbolin** [comment]
Approach looks good - one thing I was thinking was that maybe the configuration document should be a structured format (e.g. json or xml) that would allow more flexibility in how whitelisting happens.  For example, an option to whitelist by some logical check like "is Delegate" so that you wouldn't have to whitelist every variation of delegate that you use, as well as allowing Regex on Full type names for cases where types within a namespace should all be whitelisted.  But this is fine for now.

The CI build is failing on the two added tests.

**Comment by mwelsh1118** [comment]
Then tests were not completely consistent because of the static state on Helper. The last commit I added pulled the immutability-checking logic out into an ImmutableTypes class, which is constructed in the main analyzer and passed down to the individual IMM rules that need it.

Also, the handling of generics for whitelisted types was broken (`Func<int>` wasn't matching `Func<T>`). I refactored the checking a bit to no longer treat KeyValuePair and Nullable as special cases but instead include them as part of the whitelist. And I fixed the generic type handling for the configurable whitelist.

For us, I think whitelisting the Action/Func types will cover 99% of our delegate cases. So it's not a big deal to list all the variations of those.

**Comment by mwelsh1118** [comment]
Do you have any more feedback? We'd like to start taking advantage of the configurable whitelist.

**Comment by dbolin** [comment]
There's a couple things I'm not sure about - there's a single ImmutableTypes instance created and passed to each rule's Initialize method, and when the rule runs it calls `immutableTypes.Initialize(context.Compilation, context.Options, context.CancellationToken);` that seems like a possible concurrency problem.  I'd prefer to just keep passing the context to the IsImmutableType method call.  Is there a reason you didn't want to do that?

Also, ImmutableTypes holds state that tracks the current immutability status of each type it has tested.  The instance does appear to be created each time the analyzer is invoked, but do you know if we can always rely on that behavior?

File: Immutable/Apex.Analyzers.Immutable/Rules/ImmutableTypes.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace Apex.Analyzers.Immutable.Rules
+{
+    public class ImmutableTypes
```
Comment (dbolin) [inline Immutable/Apex.Analyzers.Immutable/Rules/ImmutableTypes.cs]: I'd like to keep this internal until we can move it out into a separate package

**Comment by mwelsh1118** [comment]
> I'd prefer to just keep passing the context to the IsImmutableType method call. Is there a reason you didn't want to do that?

Mainly it's simpler to avoid passing the state down the stack. Before my changes, Helper had static state that was caching ITypeSymbols, which were looked up based on the passed-in Compilation object. For example,

```csharp 
private static ITypeSymbol GetTimeSpanType(Compilation compilation)
{
    return _timeSpanSymbol 
        ?? (_timeSpanSymbol = compilation.GetTypeByMetadataName("System.TimeSpan"));
}
```
So whether it's cached in a field and referenced later or passed down on each call, the compilation is/was being treated as interchangeable across calls.

That said, I agree it would be better to avoid potentially mixing up compilations. I could refactor to call `RegisterCompilationStartAction()`, create the `ImmutableTypes` object in the callback (passing the context to the constructor), and then register the symbol action (also in the callback).  That's more or less what the BannedApiAnalyzers are doing. Let me know if that would address your concerns.

> ImmutableTypes holds state that tracks the current immutability status of each type it has tested

Wouldn't this be unchanging metadata about a given type? I'm doing it basically as an optimization to avoid recalculating the immutability status over and over.

File: Immutable/Apex.Analyzers.Immutable/Rules/ImmutableTypes.cs (hunk 1 above)
Comment (mwelsh1118) [reply Immutable/Apex.Analyzers.Immutable/Rules/ImmutableTypes.cs]: Fixed: 3b1a856c2032553b869476cc3f7e252bffb706aa

**Comment by dbolin** [comment]
> Mainly it's simpler to avoid passing the state down the stack. Before my changes, Helper had static state that was caching ITypeSymbols, which were looked up based on the passed-in Compilation object. For example,
> 
> ```cs
> private static ITypeSymbol GetTimeSpanType(Compilation compilation)
> {
>     return _timeSpanSymbol 
>         ?? (_timeSpanSymbol = compilation.GetTypeByMetadataName("System.TimeSpan"));
> }
> ```
> 
> So whether it's cached in a field and referenced later or passed down on each call, the compilation is/was being treated as interchangeable across calls.
> ...
> Wouldn't this be unchanging metadata about a given type? I'm doing it basically as an optimization to avoid recalculating the immutability status over and over.

This was only true for the built in whitelisted types that should never change regardless of any user documents/edits, etc.  Now it applies to all types.  However, currently the behavior is to create a new ImmutableTypes instance every time the analyzer runs, so it works as is.  So I'll merge this now, and if any issues come up later they can be fixed then.

**Pull request merged**